Doc/library/functools.rst

```diff
@@ -383,6 +383,28 @@ The :mod:`functools` module defines the following functions:
       The :func:`register` attribute supports using type annotations.
 
 
+.. class:: singledispatchmethod(func)
+
+   Return a new :class:`singledispatchmethod` descriptor which behaves
+   like :func:`singledispatch` except that it is designed to be used as a
+   method definition rather than being directly callable.
+
+   :class:`singledispatchmethod` can be used as follows::
+
+    class Negator:
+        @singledispatchmethod
+        def neg(self, arg):
+            raise NotImplementedError("Cannot negate a")
+
+        @neg.register
+        def _(self, arg: int):
+            return -arg
+
+        @neg.register
+        def _(self, arg: bool):
+            return not arg
+
```

Comment:
Having an example stacked with `@classmethod` would be good here. It's unfortunate that the order of application will need to be different from that when stacking `@abstractmethod` (see https://docs.python.org/3/library/abc.html#abc.abstractmethod), but we need "register" to be accessible on the resulting descriptor. So something like:

    class Negator:
        @singledispatchmethod
        @classmethod
        def neg(cls, arg):
            ...

Given the `@classmethod` example, you can then note that the same principle applies for `@staticmethod` and other decorators - to be useful, `@singledispatchmethod` typically needs to be the outermost decorator so that the `register` method is accessible.